apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,336 @@
+    */
+    function initialize(
+        ERC20 _daiToken,
+        MiniMeToken _projectToken,
+        TokenManager _projectTokenManager,
+        uint64 _vestingCliffPeriod,
+        uint64 _vestingCompletePeriod,
+        uint256 _daiFundingGoal,
+        uint256 _percentSupplyOffered,
+        uint64 _fundingPeriod,
+        address _fundraisingPool,
+        address _beneficiaryAddress,
+        uint256 _percentFundingForBenefiriary
+    )
+        external
+        onlyInit
+    {
+        require(isContract(_daiToken), ERROR_INVALID_DAI_TOKEN);
+        require(isContract(_fundraisingPool), ERROR_INVALID_POOL);
+        require(_fundingPeriod > 0, ERROR_INVALID_TIME_PERIOD);
+        require(_vestingCliffPeriod > _fundingPeriod, ERROR_INVALID_TIME_PERIOD);
+        require(_vestingCompletePeriod > _vestingCliffPeriod, ERROR_INVALID_TIME_PERIOD);
+        require(_daiFundingGoal > 0, ERROR_INVALID_DAI_FUNDING_GOAL);
+        require(_percentSupplyOffered > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentSupplyOffered < PPM, ERROR_INVALID_PERCENT_VALUE);
+        require(_beneficiaryAddress != 0x0, ERROR_INVALID_BENEFICIARY_ADDRESS);
+        require(_percentFundingForBenefiriary > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentFundingForBenefiriary < PPM, ERROR_INVALID_PERCENT_VALUE);
+
+        initialized();
+
+        daiToken = _daiToken;
+        _setProjectToken(_projectToken, _projectTokenManager);
+
+        fundraisingPool = _fundraisingPool;
+
+        vestingCliffPeriod = _vestingCliffPeriod;
+        vestingCompletePeriod = _vestingCompletePeriod;
+        fundingPeriod = _fundingPeriod;
+
+        beneficiaryAddress = _beneficiaryAddress;
+        percentFundingForBeneficiary = _percentFundingForBenefiriary;
+
+        daiFundingGoal = _daiFundingGoal;
+        percentSupplyOffered = _percentSupplyOffered;
+
+        _calculateExchangeRate();
+    }
+
+    /*
+     * Public
+     */
+
+    /**
+    * @notice Starts the sale, changing its state to Funding. After the sale is started contributors will be able to purchase project tokens.
+    */
+    function start() public auth(START_ROLE) {
+        require(currentSaleState() == SaleState.Pending, ERROR_INVALID_STATE);
+        startDate = getTimestamp64();
+        emit SaleStarted();
+    }
+
+    /**
+    * @notice Buys project tokens using the provided `_daiToSpend` dai tokens. To calculate how many project tokens will be sold for the provided, dai amount, use daiToProjectTokens(). Each purchase generates a numeric purchaseId (0, 1, 2, etc) for the caller, which can be obtained in the TokensPurchased event emitted, and is required for later refunds.
+    * @param _daiToSpend The amount of dai to spend to obtain project tokens.
+    */
+    function buy(uint256 _daiToSpend) public auth(BUY_ROLE) {
```

Comment:
Might as well make this `authP(BUY_ROLE, arr(_daiToSpend))`

Reply:
As discussed elsewhere:
It'd be interesting to limit the amount of tokens for the BUY_ROLE, but this could be circumvented easily by making multiple purchases, so will keep auth(BUY_ROLE) for now.